src/main/java/com/artipie/asto/s3/S3Storage.java

```diff
@@ -123,9 +124,21 @@ public S3Storage(final S3AsyncClient client, final String bucket) {
         ).thenApply(
             created -> new MultipartUpload(this.client, this.bucket, key, created.uploadId())
         ).thenCompose(
-            upload -> upload.upload(content)
-                .thenCompose(ignored -> upload.complete())
-        ).thenApply(response -> null);
+            upload -> upload.upload(content).handle(
+                (ignored, throwable) -> {
+                    final CompletionStage<Void> finished;
+                    if (throwable == null) {
+                        finished = upload.complete();
+                    } else {
+                        final CompletableFuture<Void> promise = new CompletableFuture<>();
+                        promise.completeExceptionally(throwable);
+                        finished = promise;
+                        upload.abort();
```

Comment:
@olegmoz it's async call, so maybe we need to use something like `return promise.thenCompose(none -> upload.abort())` here?

Reply:
@g4s8 fixed that (also tests failed because of that). Added puzzle for the problem to look into it later.